astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+        preserve_observer_frame : bool
+            If `True`, the final observer frame class will be the same as the
+            original one, and if `False` it will be the frame of the velocity
+            reference class.
+
+        Returns
+        -------
+        new_coord : `SpectralCoord`
+            The new coordinate object representing the spectral data
+            transformed based on the observer's new velocity frame.
+        """
+
+        if self.observer is None or self.target is None:
+            raise ValueError("This method can only be used if both observer "
+                             "and target are defined on the SpectralCoord.")
+
+        if hasattr(frame, 'frame'):
+            frame = frame.frame
+        elif isinstance(frame, type):
+            frame_cls = frame
+            frame = frame_cls(0 * u.m, 0 * u.m, 0 * u.m,
+                              0 * u.m / u.s, 0 * u.m / u.s, 0 * u.m / u.s,
+                              representation_type='cartesian',
+                              differential_type='cartesian')
+        elif isinstance(frame, str):
+            frame_cls = frame_transform_graph.lookup_name(frame)
+            frame = frame_cls(0 * u.m, 0 * u.m, 0 * u.m,
+                              0 * u.m / u.s, 0 * u.m / u.s, 0 * u.m / u.s,
+                              representation_type='cartesian',
+                              differential_type='cartesian')
+
+        observer = update_differentials_to_match(self.observer, frame,
+                                                 preserve_observer_frame=preserve_observer_frame)
+
+        new_coord = self._change_observer_to(observer)
+
+        return new_coord
+
+    def with_radial_velocity_shift(self, target_shift=None, observer_shift=None):
+        """
+        Apply a velocity shift to this spectral coordinate. The shift
+        can be provided as a redshift (float value) or radial velocity
+        (`~astropy.units.Quantity` with physical type of 'speed').
+
+        Parameters
+        ----------
+        target_shift : float or `~astropy.units.Quantity`
+            Shift value to apply to current target.
+        observer_shift : float or `~astropy.units.Quantity`
+            Shift value to apply to current observer.
+
+        Returns
+        -------
+        `SpectralCoord`
+            New spectral coordinate with the target/observer velocity changed
+            to incorporate the shift.
+        """
+        if observer_shift is not None and (self.target is None or
+                                           self.observer is None):
+            raise ValueError("Both an observer and target must be defined "
+                             "before applying a velocity shift.")
+
+        for arg in [x for x in [target_shift, observer_shift] if x is not None]:
+            if isinstance(arg, u.Quantity) and \
+                    arg.unit.physical_type not in ['speed', 'dimensionless']:
```

Comment:
Maybe just `args.unit.is_equivalent((u.dimensionless_unscaled, u.km/u.s))`? Allows possible user-defined equivalencies that are in play.

Reply:
Done!